esp32: cmake: fix toolchain path for espressif socs

sylvioalves opened the pull request:
Fixes toolchain include path and build warnings.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/37202

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves committed:
esp32: cmake: fix toolchain path for espressif soc

To keep bisectability, cmake and west modifications
needs to be in sync.

cmake: update external project configuration

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves commented:
@nashif This let's esp32c3 working with zephyr's riscv toolchain without espressif's toolchain.  Also extends the tweak for esp32 soc.

uLipe reviewed (approved):
LGTM

stephanosio commented on boards/riscv/esp32c3_devkitm/esp32c3_devkitm.yaml:
```diff
@@ -3,4 +3,4 @@ name: ESP32-C3
 arch: riscv
 toolchain:
-  - espressif
+  - zephyr
```
Any reason to remove `espressif` here? Does Espressif's toolchain no longer support ESP32C3?

uLipe replied:
Currently esp32c3 uses zephyr toolchain for building its application code, but the bootloader was being compiled using the IDF toolchain.

The idea is to use only the Zephyr toolchain for both and extend this to other espressif SoC.

This would remove the dependency of dealing with IDF toolchain manual setting when preparing the build setup for ESP32 and its variants.

stephanosio replied:
Fair enough.

stephanosio reviewed (commented):
Please separate the following (unrelated) patches into individual commits, so we don't have to revert everything here if one of these is found to be problematic in the future:
```
esp32c3: use default zephyr riscv toolchain

esp32: update bootloader cmake to use proper toolchain

doc: update esp32 and esp32c3 documentation
```

uLipe commented:
> @nashif This let's esp32c3 working with zephyr's riscv toolchain without espressif's toolchain.  Also extends the tweak for esp32 soc.
> 

Additionally this simplifies the setup project, because IDF toolchain path would be no longer required.

Just running the Zephyr getting started is sufficient to build and flash an application.

sylvioalves commented:
> Please separate the following (unrelated) patches into individual commits, so we don't have to revert everything here if one of these is found to be problematic in the future:
> 
> ```
> esp32c3: use default zephyr riscv toolchain
> 
> esp32: update bootloader cmake to use proper toolchain
> 
> doc: update esp32 and esp32c3 documentation
> ```

@stephanosio To keep bisectability, I can only split the documentation. Both cmake files for ESP32 and ESP32C3 needs to be in the sync with west.yml update. Do you still want me to split this in 2 commits?

sylvioalves committed:
linker: esp32c3: add thread storage area

This adds missing _end area and remove
unused definitions present in the linker script

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves committed:
dtsi: esp32c3: added missing cell defs

missing address and cell size definition causes
testing failures.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves committed:
soc: esp32c3: fix implicit declaration

replace HAL call to direct read reg function

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves committed:
doc: espressif: update espressif soc guidelines

Remove unnecessary path export for linux environment
Update serial port flashing information

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

uLipe reviewed (approved):
Still LGTM

sylvioalves commented:
@nashif, buildkite keeps failing due to submodule sync issue. It seems not to be related to CI/Twister test itself.  Same happen with other tests. Locally, the test works as expected.
Is there a way either fix it?

```
-- Initialising new submodule components/bootloader/subproject/components/micro-ecc/micro-ecc...
--
  | fatal: destination path '/var/lib/buildkite-agent/zephyr-module-cache/modules/hal/espressif/components/bootloader/subproject/components/micro-ecc/micro-ecc' already exists and is not an empty directory.
  | fatal: clone of 'https://github.com/kmackay/micro-ecc.git' into submodule path '/var/lib/buildkite-agent/zephyr-module-cache/modules/hal/espressif/components/bootloader/subproject/components/micro-ecc/micro-ecc' failed
  | Failed to clone 'components/bootloader/subproject/components/micro-ecc/micro-ecc'. Retry scheduled
  | BUG: submodule considered for cloning, doesn't need cloning any more?
  | From https://github.com/kmackay/micro-ecc
  | * branch            HEAD       -> FETCH_HEAD
  | Submodule path 'components/bootloader/subproject/components/micro-ecc/micro-ecc': checked out 'd037ec89546fad14b5c4d5456c2e23a71e554966'
  | -- Initialising new submodule components/bt/controller/lib_esp32...
  | Submodule 'components/bt/controller/lib_esp32' (https://github.com/espressif/esp32-bt-lib.git) registered for path 'components/bt/controller/lib_esp32'
  | fatal: Needed a single revision
  | Unable to find current revision in submodule path 'components/bt/controller/lib_esp32'
  | CMake Error at /var/lib/buildkite-agent/zephyr-module-cache/modules/hal/espressif/tools/cmake/git_submodules.cmake:48 (message):
  | Git submodule init failed for components/bt/controller/lib_esp32
  | Call Stack (most recent call first):
  | /var/lib/buildkite-agent/zephyr-module-cache/modules/hal/espressif/tools/cmake/build.cmake:77 (git_submodule_check)
  | /var/lib/buildkite-agent/zephyr-module-cache/modules/hal/espressif/tools/cmake/build.cmake:163 (__build_get_idf_git_revision)
  | /workdir/modules/hal/espressif/tools/cmake/idf.cmake:50 (__build_init)
  | /workdir/modules/hal/espressif/tools/cmake/project.cmake:7 (include)
  | CMakeLists.txt:34 (include)
```

sylvioalves commented:
@cfriedt, do you know how to fix this buildkite issue? Once it starts building the project, it tries to `git submodule` but fails, which causes build failure.

cfriedt commented:
@sylvioalves - I wonder if this is the same root cause as the qemu_riscv32 test failure that I disabled on Friday.

I don't have a solution for it currently in CI but I can bring it up in the release meeting.

sylvioalves commented:
> @sylvioalves - I wonder if this is the same root cause as the qemu_riscv32 test failure that I disabled on Friday.
> 
> I don't have a solution for it currently in CI but I can bring it up in the release meeting.

Is there a way to disable CI tests for this board? I have tried increasing the ignore list but it seems its lacking a few tests.

cfriedt commented:
> @cfriedt Is there a way to disable CI tests for this board? I have tried increasing the ignore list but it seems its lacking a few tests.

You could add a `platform_exclude: <board name>` to the relevant `testcase.yaml` files, but that should really only be done in rare cases.

sylvioalves commented:
> @sylvioalves - I wonder if this is the same root cause as the qemu_riscv32 test failure that I disabled on Friday.
> 
> I don't have a solution for it currently in CI but I can bring it up in the release meeting.

@cfriedt, when is it supposed to happen?

cfriedt commented:
> @cfriedt, when is it supposed to happen?

It happened already. See my earlier message about how to disable the testcase for your board. You can make that change in this PR.

sylvioalves commented:
> > @cfriedt, when is it supposed to happen?
> 
> It happened already. See my earlier message about how to disable the testcase for your board. You can make that change in this PR.

I see. However, that action won't be successful. 
The error is related to `git submodule` call. And it is triggered if any test is performed with this target board. So it is not related to a specific test itself.  
I have tried disabling almost all tests by adding that in the `ignore-tags` as you can below, however, this list does not contain all `tags`.

```c
arch: riscv
toolchain:
  - zephyr
testing:
  ignore_tags:
    - bluetooth
    - net
    - posix
    - crypto
    - logging
    - shell
    - jwt
    - cpp
    - kernel
    - test_framework
    - drivers
    - cmsis_rtos
 ```  
So I see 2 ways of handling this:  
1) excluding the board from all tests (not sure how to do that)
2) fixing `git submodule` issue, either on `hal_espressif` repository or in CI itself. But I was not able to understand the problem yet as `git sumodule` keeps failing.

Hope you can address it someway.  
Thanks!

cfriedt commented:
@sylvioalves - can you either copy & paste the error you are seeing or share a link to it?

The only failure I see in CI is in `tests/subsys/pm/device_runtime/subsys.pm.device_pm`.

sylvioalves commented:
> @sylvioalves - can you either copy & paste the error you are seeing or share a link to it?
> 
> The only failure I see in CI is in `tests/subsys/pm/device_runtime/subsys.pm.device_pm`.

Find it below:
https://buildkite.com/zephyr/zephyr/builds/36113#8072b5e4-cba8-45a7-a096-d2816f01bf4b/918-1097

cfriedt commented:
@tejlmand - could you investigate why this build is failing? It seems to be that git submodule checkout does not work.

Please create a separate issue if there is a bug and we'll refer to this PR Q

sylvioalves commented:
> @tejlmand - could you investigate why this build is failing? It seems to be that git submodule checkout does not work.
> 
> Please create a separate issue if there is a bug and we'll refer to this PR.

@tejlmand, did you have time to check it?

cfriedt commented:
@galak - any idea why git submodule checkout might not work in CI?

The git submodule part works for me locally with success as well.
```
-- Found Git: /opt/homebrew/bin/git (found version "2.30.2") 
-- Initialising new submodule components/bootloader/subproject/components/micro-ecc/micro-ecc...
Submodule 'components/bootloader/subproject/components/micro-ecc/micro-ecc' (https://github.com/kmackay/micro-ecc.git) registered for path 'components/bootloader/subproject/components/micro-ecc/micro-ecc'
Cloning into '/Users/cfriedt/workspace/zephyrproject/modules/hal/espressif/components/bootloader/subproject/components/micro-ecc/micro-ecc'...
Submodule path 'components/bootloader/subproject/components/micro-ecc/micro-ecc': checked out 'd037ec89546fad14b5c4d5456c2e23a71e554966'
-- Initialising new submodule components/bt/controller/lib_esp32...
Submodule 'components/bt/controller/lib_esp32' (https://github.com/espressif/esp32-bt-lib.git) registered for path 'components/bt/controller/lib_esp32'
Cloning into '/Users/cfriedt/workspace/zephyrproject/modules/hal/espressif/components/bt/controller/lib_esp32'...
Submodule path 'components/bt/controller/lib_esp32': checked out 'ec61ca3caa64874e11f39a92654b4160bb5db06e'
-- Initialising new submodule components/esp_wifi/lib...
Submodule 'components/esp_wifi/lib' (https://github.com/espressif/esp32-wifi-lib.git) registered for path 'components/esp_wifi/lib'
Cloning into '/Users/cfriedt/workspace/zephyrproject/modules/hal/espressif/components/esp_wifi/lib'...
Submodule path 'components/esp_wifi/lib': checked out '6e096b5e7a8b527732499f943c70d3203f2abd2f'
-- Initialising new submodule components/esptool_py/esptool...
Submodule 'components/esptool_py/esptool' (https://github.com/espressif/esptool.git) registered for path 'components/esptool_py/esptool'
Cloning into '/Users/cfriedt/workspace/zephyrproject/modules/hal/espressif/components/esptool_py/esptool'...
Submodule path 'components/esptool_py/esptool': checked out '4698b396730b23fb4aab023c5fb1744db957fc4c'
```

sylvioalves commented:
@cfriedt @galak Hi, any updates?

galak commented on west.yml:
```diff
@@ -64,11 +64,7 @@ manifest:
       revision: f525ee6f468c2c68119731e1453ad72d189a1277
       path: modules/hal/espressif
-      groups:
-        - hal
```
why the removal here of `hal` group?

sylvioalves replied:
In fact, I just fixed it because some previous PR messed with it. It has duplicated entries:

```c
    - name: hal_espressif
      revision: f525ee6f468c2c68119731e1453ad72d189a1277
      path: modules/hal/espressif
      groups:
        - hal
      west-commands: west/west-commands.yml
      revision: c21d252a2188fcb60d52dceb6b018a9f4a371817
      path: modules/hal/espressif
      groups:
        - hal
```

galak commented:
Did west add support for handling `git submodules`

Any thoughts @mbolivar-nordic

sylvioalves commented:
> Did west add support for handling `git submodules`
> 
> Any thoughts @mbolivar-nordic

So, I added the `submodules: true` into `west.yml`. Althought it works locally, it seems CI is not retrieving the submodules correctly as the error below complains about missing includes:

```c
/workdir/modules/hal/espressif/components/bootloader/subproject/components/micro-ecc/uECC_verify_antifault.c:10:10: fatal error: micro-ecc/uECC.c: No such file or directory
--
  | 10 \| #include "micro-ecc/uECC.c"
  | \|          ^~~~~~~~~~~~~~~~~~
  | compilation terminated.
  | [21/81] Building C object esp-idf/soc/CMakeFiles/__idf_soc.dir/esp32c3/timer_periph.c.obj
```

mbolivar-nordic commented:
> So, I added the `submodules: true` into `west.yml`. Although it works locally, it seems CI is not retrieving the submodules correctly as it complains about missing includes:

I'll leave that to you all to figure out then since it's working locally. I'm not aware of any issues within west itself. Nordic uses submodule support in our west.yml file so it does work.

galak commented:
Marked DNM until we are clear about use of `submodules` in west.yaml

sylvioalves commented:
> Marked DNM until we are clear about use of `submodules` in west.yaml

@galak, it works locally. Who can check that in CI?

sylvioalves commented:
@galak I updated this PR to keep esp-idf default toolchain instead of Zephyr's toolchain. This fixes CI issue for the moment. We can now merge this. I'll open a new PR eventually to use Zephyr's riscv toolchain and then we can work out the CI issue.

galak commented:
> @galak I updated this PR to keep esp-idf default toolchain instead of Zephyr's toolchain. This fixes CI issue for the moment. We can now merge this. I'll open a new PR eventually to use Zephyr's riscv toolchain and then we can work out the CI issue.

We need clarity on the `submodule` usage in west.yaml.

@carlescufi @nashif - are we ok with this?

sylvioalves commented:
> > @galak I updated this PR to keep esp-idf default toolchain instead of Zephyr's toolchain. This fixes CI issue for the moment. We can now merge this. I'll open a new PR eventually to use Zephyr's riscv toolchain and then we can work out the CI issue.
> 
> We need clarity on the `submodule` usage in west.yaml.
> 
> @carlescufi @nashif - are we ok with this?

Adding the `submodule: true` causes the tool to retrieve and/or update 4 different submodules required for esp32:  
* wifi libraries
* bluetooth libraries
* bootloader microecc sources
* esptool sources

Currently we use `west espressif update` to perform that action. If adding `submodules` to west.yml is not the correct approach, I can revert that.

carlescufi reviewed (dismissed):
I just realized that the submodules are actually pulling binaries too, and we haven't reached a conclusion about those yet in the TSC. Can you please split this up into 2 PRs? One with the submodule change and another one with the rest of the changes that can go in immediately?

sylvioalves commented:
> I just realized that the submodules are actually pulling binaries too, and we haven't reached a conclusion about those yet in the TSC. Can you please split this up into 2 PRs? One with the submodule change and another one with the rest of the changes that can go in immediately?

sure!

sylvioalves commented:
@carlescufi @galak submodules removed from this PR.

sylvioalves commented:
@carlescufi can we merge this?

sylvioalves commented:
Just rebased with main to fix conflicts. @cfriedt PTAL.

cfriedt reviewed (commented):
LGTM - thanks for filling bring a workaround for that CI issue

sylvioalves commented:
@cfriedt Can we merge this?
